### PR TITLE
gnrc: remove circular module dependencies

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
According to my analysis (which might be incomplete :sweat_smile:) this removes all circles within GNRC's module dependencies.

### Testing procedure
`tests/gnrc_upd` should still work as expected, including when sending fragmented IPv6 packets (see release specs). The modules of which the dependencies were changed should still compile when isolated:

```
USEMDODULE=gnrc_icmpv6 make -C examples/hello-world
USEMDODULE=gnrc_ipv6_ext make -C examples/hello-world
```

### Issues/PRs references
Finalizes the hunt for circular dependencies in GNRC, so related to Kconfig migration, phase 2.

### CODE REVIEW

**Comment by miri64** [comment]
If there is interest, I can provide my (prototype) tool as a checker for circular dependencies within the make system.

**Comment by fjmolinas** [comment]
> If there is interest, I can provide my (prototype) tool as a checker for circular dependencies within the make system.

I'm interested!

**Comment by leandrolanzieri** [comment]
@miri64 the CI shows some issues of what it seems missing modules. Could you take a look?

**Comment by miri64** [comment]
> @miri64 the CI shows some issues of what it seems missing modules. Could you take a look?

Should be fixed now.

**Comment by miri64** [comment]
Still found some dependencies.

**Comment by fjmolinas** [comment]
@miri64 I used your script on `examples/suit_update` and it found?

```
Found a cyclic dependencies:
 - sock_udp -> gnrc_sock_udp
 - gnrc_sock_udp -> sock_udp
```

might there be something missing?

**Comment by fjmolinas** [comment]
Found another one (unless I'm misusing your tool):

```
Found a cyclic dependencies:
 - gnrc_sixlowpan_router -> gnrc_ipv6_router
 - gnrc_ipv6_router -> gnrc_sixlowpan_router
```

**Comment by miri64** [comment]
> Still found some dependencies.

@fjmolinas thats what I meant by that ;-)

**Comment by fjmolinas** [comment]
>  @fjmolinas thats what I meant by that ;-)

Yes sorry, wrote the comment before refreshing the page, so i hadn't seen your comment :)

**Comment by miri64** [comment]
> @miri64 I used your script on `examples/suit_update` and it found?
> 
> ```
> Found a cyclic dependencies:
>  - sock_udp -> gnrc_sock_udp
>  - gnrc_sock_udp -> sock_udp
> ```
> 
> might there be something missing?

This would effectively revert https://github.com/RIOT-OS/RIOT/pull/12931. Maybe we should discuss this first :-/

**Comment by miri64** [comment]
Explanations why I removed a certain dependency can be found in the corresponding commit. As I said, lets discuss

```diff
diff --git a/Makefile.dep b/Makefile.dep
index 415130306b..e93361d627 100644
--- a/Makefile.dep
+++ b/Makefile.dep
@@ -639,9 +639,6 @@ ifneq (,$(filter gnrc,$(USEMODULE)))
   USEMODULE += gnrc_netif
   USEMODULE += gnrc_netif_hdr
   USEMODULE += gnrc_pktbuf
-  ifneq (,$(filter sock_udp, $(USEMODULE)))
-    USEMODULE += gnrc_sock_udp
-  endif
 endif
 
 ifneq (,$(filter gnrc_pktbuf, $(USEMODULE)))
```

first. As this PR is about removing cyclic dependencies in GNRC, I leave the `stdio_uart` => `stdio_uart_rx` => `stdio_uart` issue for someone else to ponder on.

**Comment by fjmolinas** [comment]
> Explanations why I removed a certain dependency can be found in the corresponding commit. As I said, lets discuss

So, `lwip_sock_udp`, `gnrc_sock_udp` , `etc_sock_udp` seem like implementations of `sock_udp`. It seems that what one should do is:

```
USEMODULE += sock_udp
```

And then select the implementation (which could have a default value)

```
ifeq (, $(filter %_sock_udp, $(USEMODULE)))
  ifneq (, $(fiter gnrc_%, $(USEMODULE)))
    USEMODULE += gnrc_sock_udp
  endif
  ...
endif
```

And therefore I would do this change

```diff
diff --git a/Makefile.dep b/Makefile.dep
index 0597c08a18..cc3dead871 100644
--- a/Makefile.dep
+++ b/Makefile.dep
@@ -154,7 +154,6 @@ endif
 ifneq (,$(filter gnrc_sock_udp,$(USEMODULE)))
   USEMODULE += gnrc_udp
   USEMODULE += random     # to generate random ports
-  USEMODULE += sock_udp
 endif
 
 ifneq (,$(filter gnrc_sock,$(USEMODULE)))
```

Which means:

- `sock_udp` must be included
- a default implementation of `sock_udp` will be selected if none is selected explicitly.

**Comment by fjmolinas** [comment]
> Which means:
> 
> - `sock_udp` must be included
> - a default implementation of `sock_udp` will be selected if none is selected explicitly.

Does this make sense?

**Comment by miri64** [comment]
>     * a default implementation of `sock_udp` will be selected if none is selected explicitly.

That's just it, there is no "default" implementation. `sock` is always an implementation by the stack. But your argument makes sense. Picking just `sock` + network stack is sounds more sensible, however, that would change the current status quo (applications picking the specific `sock` implementation.. just grep for `gnrc_sock_udp` ;-))

**Comment by fjmolinas** [comment]
> That's just it, there is no "default" implementation. sock is always an implementation by the stack. But your argument makes sense. Picking just sock + network stack is sounds more sensible, however, that would change the current status quo (applications picking the specific sock implementation.. 

Then maybe no default should be selected, but just the one matching the stack that is in use. If none is, then show an error message.

> just grep for gnrc_sock_udp ;-))

Yes I know hahaha, had done just that. But I think that when we address these circular dependencies, we will have to change some of the status quo stuff. I think it generally makes sense to have the the "api" (`sock_udp`) select the "implementation" (`%_sock_udp`). I had similar thoughts on a PR regarding `xtimer` and `xtimer_on_ztimer`, https://github.com/RIOT-OS/RIOT/pull/14123#issuecomment-649056129.

@leandrolanzieri any opinions here?

**Comment by miri64** [comment]
> Then maybe no default should be selected, but just the one matching the stack that is in use. If none is, then show an error message.

The `sock_*` pseudo-modules are not selected, they are pulled in for dependency resolution. E.g. for `posix_socket`.

**Comment by leandrolanzieri** [comment]
> I think it generally makes sense to have the the "api" (sock_udp) select the "implementation" (%_sock_udp).

I tend to agree with this. Thinking from the point of view of modelling this in Kconfig, the API could have it's own symbol and model a choice between all the implementations. This is because they should be mutually exclusive.

**Comment by miri64** [comment]
> Yes I know hahaha, had done just that. But I think that when we address these circular dependencies, we will have to change some of the status quo stuff.

Ok, but then I would rather change the status quo in a separate PR, so it doesn't happen on the back street ;-).

**Comment by miri64** [comment]
See https://github.com/RIOT-OS/RIOT/pull/14906

**Comment by leandrolanzieri** [comment]
Murdock spotted a couple more of issues here

**Comment by miri64** [comment]
Should be fixed now.

**Comment by leandrolanzieri** [comment]
The NDP test [is failing](https://ci.riot-os.org/RIOT-OS/RIOT/14886/76a462296e42b60b3450e444ada7ce55b4195d40/output/compile/tests/gnrc_ndp/native:gnu.txt)

**Comment by miri64** [comment]
> The NDP test [is failing](https://ci.riot-os.org/RIOT-OS/RIOT/14886/76a462296e42b60b3450e444ada7ce55b4195d40/output/compile/tests/gnrc_ndp/native:gnu.txt)

already better than not compiling :sweat_smile:

**Comment by miri64** [comment]
It passes now again.

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 1]
```diff
@@ -588,16 +588,16 @@ int gnrc_netif_ipv6_addr_add_internal(gnrc_netif_t *netif,
 
         gnrc_netif_ipv6_bus_post(netif, GNRC_IPV6_EVENT_ADDR_VALID, &netif->ipv6.addrs[idx]);
     }
-#if IS_ACTIVE(CONFIG_GNRC_IPV6_NIB_SLAAC)
-    else if (!gnrc_netif_is_6ln(netif)) {
+    else if (IS_USED(MODULE_GNRC_IPV6) &&
```
Comment (leandrolanzieri) [inline sys/net/gnrc/netif/gnrc_netif.c:591]: Is checking for `MODULE_GNRC_IPV6` necessary? Wouldn't IPv6 support be always enabled when using `MODULE_GNRC_IPV6_NIB`?

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:591]: That would reintroduce a circular dependency again, as `gnrc_ipv6` requires the `gnrc_ipv6_nib` module.

Comment (leandrolanzieri) [reply sys/net/gnrc/netif/gnrc_netif.c:591]: Ah yeah. Thanks for clarifying.

**Comment by leandrolanzieri** [comment]
I've run the release spec UDP and ICMPv6 tests and everything looks good. Also tried compiling `gnrc_ipv6_ext` and `gnrc_icmpv6` on their own as suggested in the description and works. From my side I think this could be squashed, @fjmolinas ?

**Comment by fjmolinas** [comment]
> I've run the release spec UDP and ICMPv6 tests and everything looks good. Also tried compiling gnrc_ipv6_ext and gnrc_icmpv6 on their own as suggested in the description and works. From my side I think this could be squashed, @fjmolinas ?

No comments either, go ahead and squash.

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
Murdock is happy now. Still needs an ACK.

**Review by fjmolinas — APPROVED** [review]
ACK, I tested for `tests/gnrc_udp` `examples/gnrc_networking` the selected modules had not changed. I also ran the cyclic dependency script in #14900, only `sock` dependencies are showing up.

**Comment by fjmolinas** [comment]
GO!

**Comment by miri64** [comment]
> I also ran the cyclic dependency script in #14900, only `sock` dependencies are showing up.

That's no guarantee that there are no further cyclic dependencies, as it only shows the first it finds, but last time I checked only the `sock` and `stdio_read`-`stdio` cycle remained.

**Pull request merged**